### PR TITLE
[NO REVIEW] Update the CPUID and XSAVE logics for APX

### PR DESCRIPTION
**Pull request by Ruihan-Yin**
### [No need to review this PR for now]

Overview on the changes:

1. Infrastructural changes on `XArchIntrinsicConstants`: Compress all the Avx512 related flags into 1 - `Avx512f+bw+cd+dq+vl` to `Avx512`, this saves more space in `XArchIntrinsicConstants` so that we can hold more x86 ISAs, like here, APX.
- Edit: #103144 and #103241 have resolved this part.
3. CPUID check updates for APX: `CR4[XSAVE]`(existing) -> `XCR0[APX_F]` -> `CPUID(7,1).EDX[APX_F]` - the current status is that due to the missing macro definition for APX on the OS level, the second check will fail anyways, and it may break the build on CI (to be verified). 
4. APX introduced 16 EGPRs, which are XSTATE enabled, this PR provides the corresponding changes for XSAVE behaviors.

### CODE REVIEW

**Comment by tannergooding** [comment]
I've gone ahead and marked this as `NO-MERGE` for now. If it is ready for merge prior to `main` opening for .NET 10 changes, we would want to go through some additional considerations as to the impact it would have on .NET 9 and whether or not we want it excluded from the .NET 9 build such as via an `#if defined(FEATURE_APX)` flag.

**Comment by Ruihan-Yin** [comment]
> I've gone ahead and marked this as `NO-MERGE` for now. If it is ready for merge prior to `main` opening for .NET 10 changes, we would want to go through some additional considerations as to the impact it would have on .NET 9 and whether or not we want it excluded from the .NET 9 build such as via an `#if defined(FEATURE_APX)` flag.

Yes, please mark it as `NO-MERGE`, thanks.

File: apx-doc/README.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# APX Integration in .NET
+
+Let's keep documentation on APX integration and notes on things here. I will evolve this as necessary.
```
Comment (tannergooding) [inline apx-doc/README.md:3]: We should move this under `docs/design/features/xarch-apx.md` or maybe under `docs/design/coreclr/jit/xarch-apx.md`, just to keep it in the same general place as our other docs.

I like the idea of having a documentation covering how the feature is used and integrates with the rest of the JIT.

File: src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs [hunk 2]
```diff
@@ -70,20 +70,14 @@ private static class XArchIntrinsicConstants
...
-            public const int Avx512cd_vl = 0x100000;
-            public const int Avx512dq = 0x200000;
-            public const int Avx512dq_vl = 0x400000;
-            public const int Avx512Vbmi = 0x800000;
-            public const int Avx512Vbmi_vl = 0x1000000;
-            public const int Serialize = 0x2000000;
-            public const int Avx10v1 = 0x4000000;
-            public const int Avx10v1_v256 = 0x8000000;
-            public const int Avx10v1_v512 = 0x10000000;
+            public const int Avx512 = 0x8000;
```
Comment (jkotas) [inline src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs:73]: If you would like to make these bits more compact, it would be better to submit it as separate PR. There are free bits now so you do not need it to add the Apx bit.

It may be better to avoid folding the features together at the PAL level:
```
            public const int Avx512f = ...;
            public const int Avx512f_vl = ...;
            public const int Avx512bw = ...;
            public const int Avx512cd = ...;
            public const int Avx512dq = ...;
            public const int Avx512Vbmi = ...;
```
I understand that it does not save as many bits, but it is cleaner (avoids applying JIT policy at the PAL level).

File: src/coreclr/gc/vxsort/isa_detection.cpp [hunk 3]
```diff
@@ -66,36 +78,44 @@ SupportedISA DetermineSupportedISA()
...
+
+    if (__builtin_cpu_supports("avx512"))
     {
-        return SupportedISA::None;
+        IsaFlags |= (int)SupportedISA::APX;
     }
+
+    if (__builtin_cpu_supports("avx2"))
+    {
+        IsaFlags |= (int)SupportedISA::APX;
```
Comment (jkotas) [inline src/coreclr/gc/vxsort/isa_detection.cpp:115]: It is not necessary to detect APX here until the APX-specific implementation of the GC sort gets added.

File: src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs (hunk 2 above)
Comment (Ruihan-Yin) [reply src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs:73]: Thanks for the inputs, I can work on this part soon

File: src/coreclr/inc/corinfoinstructionset.h [hunk 4]
```diff
@@ -78,41 +78,43 @@ enum CORINFO_InstructionSet
...
-    InstructionSet_AVX512CD_X64=62,
-    InstructionSet_AVX512CD_VL_X64=63,
-    InstructionSet_AVX512DQ_X64=64,
-    InstructionSet_AVX512DQ_VL_X64=65,
-    InstructionSet_AVX512VBMI_X64=66,
-    InstructionSet_AVX512VBMI_VL_X64=67,
-    InstructionSet_AVX10v1_X64=68,
-    InstructionSet_AVX10v1_V256_X64=69,
-    InstructionSet_AVX10v1_V512_X64=70,
+    InstructionSet_APX=36,
```
Comment (huoyaoyuan) [inline src/coreclr/inc/corinfoinstructionset.h:81]: Does APX support 32bit at all?

> Intel®Advanced Performance Extensions (Intel® APX) expands the Intel® 64 instruction set

or is everything just generated with both x86 and x64?

File: src/coreclr/pal/src/include/pal/context.h [hunk 5]
```diff
@@ -384,9 +384,9 @@ bool Xstate_IsApxSupported();
 #define XFEATURE_MASK_AVX512 (XFEATURE_MASK_OPMASK | XFEATURE_MASK_ZMM_Hi256 | XFEATURE_MASK_Hi16_ZMM)
 #endif // XFEATURE_MASK_AVX512
 
-#ifndef XFEATURE_MASK_APX
-#define XFEATURE_MASK_APX (1 << XSTATE_APX)
-#endif // XFEATURE_MASK_APX
+// TODO-xarch-apx: the definition of XSTATE mask value for APX is now missing on the OS level, 
+//                 we are currently using bare value to hack it through the build process, and test the implementation through CI.
+//                 those changes will be removed when we have the OS support for APX.
```
Comment (tannergooding) [inline src/coreclr/pal/src/include/pal/context.h:389]: Why not do this via the following so that things can still function as expected, even if built on a downlevel OS without the support?
```cpp
#ifndef XSTATE_APX
#define XSTATE_APX 19
#endif
```

File: src/coreclr/inc/corinfoinstructionset.h (hunk 4 above)
Comment (Ruihan-Yin) [reply src/coreclr/inc/corinfoinstructionset.h:81]: Thanks for bringing this up,

No, APX features is available only under 64-bit mode
Reference: https://www.intel.com/content/www/us/en/content-details/819797/intel-advanced-performance-extensions-intel-apx-architecture-specification.html
>3.1.4 Intel® APX features are only available in IA-32e 64-bit Protected Mode, and are an XSAVE-enabled feature
which requires XCR0 enabling before using the new Intel® APX ISA, new Intel® APX prefixes (REX2) and
prefix extensions (EVEX extensions). See section 3.1.4.2 for details on XCR0-enabling for Intel® APX

Based on this fact, I understand APX could be different from existing ISAs and we are open to take suggestions on this part.

File: src/coreclr/pal/src/include/pal/context.h (hunk 5 above)
Comment (Ruihan-Yin) [reply src/coreclr/pal/src/include/pal/context.h:389]: I tried this way locally on my own dev machine with Windows OS, the build will fail with error report saying the macros are not defined, and it seems that, at least on Windows, `XSTATE_APX` and `XSTATE_MASK_APX` shall be defined in `winnt.h` (I am not 100% sure about the reason, is it because we are using Windows API in the CPUID checks?), I can fix the error by adding the definition there, but it might not be the solution in CI environment.

**Comment by Ruihan-Yin** [comment]
Hi @tannergooding 

Since the `XArchIntrinsicConstant` has been updated, I will start to work on this PR again, can we reopen this PR? Thanks!

**Comment by tannergooding** [comment]
> I will start to work on this PR again, can we reopen this PR? Thanks!

GitHub blocks it being reopened due to the branch having been force-pushed or recreated since the time the PR was closed
image

**Comment by Ruihan-Yin** [comment]
> > I will start to work on this PR again, can we reopen this PR? Thanks!
> 
> GitHub blocks it being reopened due to the branch having been force-pushed or recreated since the time the PR was closed image

I updated my branch beforehand, my bad, I will create another draft PR.
